cargo-workspaces/src/publish.rs

```diff
@@ -85,60 +85,52 @@ impl Publish {
         let (names, visited) = dag(&pkgs);
 
         // Filter out private packages
-        let visited = visited
-            .into_iter()
-            .filter(|x| {
-                if let Some((pkg, _)) = pkgs.iter().find(|(p, _)| p.manifest_path == *x) {
-                    return pkg.publish.is_none()
-                        || !pkg.publish.as_ref().expect(INTERNAL_ERR).is_empty();
-                }
+        let visited = filter_private(visited, &pkgs);
 
-                false
-            })
-            .collect::<Set<_>>();
-
-        let http_client = create_http_client(&metadata.workspace_root, &self.token)?;
+        let http_client = create_http_client(&metadata.workspace_root, &self.registry.token)?;
 
         for p in &visited {
             let (pkg, version) = names.get(p).expect(INTERNAL_ERR);
             let name = pkg.name.clone();
+
+            if self.dry_run {
+                info!("checking", name);
+
+                if !self.no_verify && !self.build(&metadata.workspace_root, p)? {
+                    warn!("build failed", "");
+                }
+
+                basic_checks(pkg)?;
+            }
+
             let mut args = vec!["publish"];
 
             let name_ver = format!("{} v{}", name, version);
-
-            let index_url = if let Some(registry) = self
-                .registry
-                .as_ref()
-                .or_else(|| pkg.publish.as_deref().and_then(|x| x.get(0)))
-            {
-                let registry_url = cargo_config_get(
-                    &metadata.workspace_root,
-                    &format!("registries.{}.index", registry),
-                )?;
-                IndexUrl::NonCratesIo(registry_url.into())
-            } else {
-                IndexUrl::crates_io(None, None, None)?
-            };
+            let index_url = package_registry(&metadata, self.registry.registry.as_ref(), pkg)?;
 
             if is_published(&http_client, index_url, &name, version)? {
                 info!("already published", name_ver);
                 continue;
             }
 
-            if self.no_verify {
+            if self.dry_run {
+                args.push("--dry-run");
+            }
```

Comment:
Do you think it will work? When I checked before, `cargo publish --dry-run` would fail on the 2nd package, because we didn't publish the previous one. This is why I didn't supply this argument before ([as mentioned here](https://github.com/pksunkara/cargo-workspaces/pull/154#discussion_r1672015096)).

Probably this should be removed.

Reply:
Just checked, the command currently doesn't work. See an example output when trying `--dry-run` with `fixtures/normal`:

```
error: failed to prepare local package for uploading

Caused by:
  no matching package named `dep1` found
  location searched: registry `crates-io`
  required by package `top v0.1.0 (/home/popzxc/workspace/cargo-workspaces/fixtures/normal/top)`
```

I will prepare a fix and add tests.

Reply:
The `--no-verify` was added when `dry-run` is enabled and that would make sure that dependencies are not checked for being published. This is how the other tools in this space are doing it too.

It was working for me in Cargo 1.75. I had a monorepo with half-published crates and all of it worked.

Reply:
Oh, sorry, you're absolutely right! I forgot to reinstall the binary when checking 🤦 
Will revert corresponding changes in #157